sklearn/linear_model/tests/test_base.py

```diff
@@ -324,6 +324,67 @@ def test_csr_preprocess_data():
     assert_equal(csr_.getformat(), 'csr')
 
 
+def test_dtype_preprocess_data():
+    n_samples = 200
+    n_features = 2
+    X = rng.rand(n_samples, n_features)
+    y = rng.rand(n_samples)
+
+    X_32 = np.asarray(X, dtype=np.float32)
+    y_32 = np.asarray(y, dtype=np.float32)
+    X_64 = np.asarray(X, dtype=np.float64)
+    y_64 = np.asarray(y, dtype=np.float64)
+
+    for fit_intercept in [True, False]:
+        for normalize in [True, False]:
+
+            Xt_32, yt_32, X_mean_32, y_mean_32, X_norm_32 = \
+                _preprocess_data(X_32, y_32, fit_intercept=fit_intercept,
+                                 normalize=normalize, return_mean=True)
+
+            Xt_64, yt_64, X_mean_64, y_mean_64, X_norm_64 = \
+                _preprocess_data(X_64, y_64, fit_intercept=fit_intercept,
+                                 normalize=normalize, return_mean=True)
+
+            Xt_3264, yt_3264, X_mean_3264, y_mean_3264, X_norm_3264 = \
+                _preprocess_data(X_32, y_64, fit_intercept=fit_intercept,
+                                 normalize=normalize, return_mean=True)
+
+            Xt_6432, yt_6432, X_mean_6432, y_mean_6432, X_norm_6432 = \
+                _preprocess_data(X_64, y_32, fit_intercept=fit_intercept,
+                                 normalize=normalize, return_mean=True)
+
+            assert_equal(Xt_32.dtype, np.float32)
+            assert_equal(yt_32.dtype, np.float32)
+            assert_equal(X_mean_32.dtype, np.float32)
+            assert_equal(y_mean_32.dtype, np.float32)
+            assert_equal(X_norm_32.dtype, np.float32)
+
+            assert_equal(Xt_64.dtype, np.float64)
+            assert_equal(yt_64.dtype, np.float64)
+            assert_equal(X_mean_64.dtype, np.float64)
+            assert_equal(y_mean_64.dtype, np.float64)
+            assert_equal(X_norm_64.dtype, np.float64)
+
+            assert_equal(Xt_3264.dtype, np.float32)
+            assert_equal(yt_3264.dtype, np.float32)
+            assert_equal(X_mean_3264.dtype, np.float32)
+            assert_equal(y_mean_3264.dtype, np.float32)
+            assert_equal(X_norm_3264.dtype, np.float32)
+
+            assert_equal(Xt_6432.dtype, np.float64)
+            assert_equal(yt_6432.dtype, np.float64)
+            assert_equal(X_mean_6432.dtype, np.float64)
+            assert_equal(y_mean_6432.dtype, np.float64)
+            assert_equal(X_norm_6432.dtype, np.float64)
+
+            assert_array_almost_equal(Xt_32, Xt_64)
```

Comment:
But doing `assert_array_equal(X_32, X_32_initial)` I don't know if the dtype is properly tested...